symbolic-syntax/src/Data/Macaw/Symbolic/Syntax.hs

```diff
@@ -0,0 +1,638 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+-- | 'LCSC.ParserHooks' for parsing macaw-symbolic CFGs. Originally ported from:
+--
+-- https://github.com/GaloisInc/ambient-verifier/blob/eab04abb9750825a25ec0cbe0379add63f05f6c6/src/Ambient/FunctionOverride/Extension.hs#L863
+module Data.Macaw.Symbolic.Syntax (
+    TypeAlias(..)
+  , TypeLookup(..)
+  , x86_64LinuxTypes
+  , ExtensionParser
+  , SomeExtensionWrapper(..)
+  , extensionParser
+  , extensionTypeParser
+  , machineCodeParserHooks
+  ) where
+
+import Prelude
+
+import           Control.Applicative ( Alternative(empty) )
+import           Control.Monad.IO.Class ( MonadIO(..) )
+import           Control.Monad.State.Class ( MonadState )
+import           Control.Monad.Writer.Class ( MonadWriter )
+import qualified Data.Map.Strict as Map
+import qualified Data.Parameterized.Context as Ctx
+import qualified Data.Parameterized.NatRepr as PN
+import           Data.Parameterized.Some ( Some(..) )
+import qualified Data.Text as DT
+import qualified Data.Vector as DV
+import           GHC.TypeNats ( KnownNat, type (<=) )
+import           Numeric.Natural ( Natural )
+
+import qualified Data.Macaw.CFG as DMC
+import qualified Data.Macaw.Memory as DMM
+import qualified Data.Macaw.Symbolic as DMS
+import qualified Lang.Crucible.CFG.Expr as LCE
+import qualified Lang.Crucible.CFG.Extension as LCCE
+import qualified Lang.Crucible.CFG.Reg as LCCR
+import qualified Lang.Crucible.LLVM.MemModel as LCLM
+import qualified Lang.Crucible.Syntax.Atoms as LCSA
+import qualified Lang.Crucible.Syntax.Concrete as LCSC
+import qualified Lang.Crucible.Syntax.ExprParse as LCSE
+import qualified Lang.Crucible.Types as LCT
+
+import qualified What4.Interface as WI
+import qualified What4.ProgramLoc as WP
+
+-- | The additional types beyond those built into crucible-syntax.
+data TypeAlias = Byte | Int | Long | PidT | Pointer | Short | SizeT | UidT
+  deriving (Show, Eq, Enum, Bounded)
+
+-- | Lookup function from a 'TypeAlias' to the underlying crucible type it
+-- represents.
+newtype TypeLookup = TypeLookup (TypeAlias -> (Some LCT.TypeRepr))
+
+-- | A lookup function from 'TypeAlias' to types with the appropriate width
+-- on X86_64 Linux.
+x86_64LinuxTypes :: TypeLookup
+x86_64LinuxTypes = TypeLookup $ \tp ->
```

Comment:
We should add `TypeLookup`s for more architectures than just x86-64. At the minimum, adding one for AArch32 would be nice.

Reply:
I propose upstreaming this logic into crucible-llvm-syntax: https://github.com/GaloisInc/crucible/pull/1120